Refactor reviewing and submitting incomplete application

paulrobertlloyd opened the pull request:
* Fixes review page not being viewable
* Fixes review page not being submittable
* Fixes issue where trying to complete an incomplete GCSE or Degree from the review page, wouldn’t allow you to answer the follow on questions about year, grade etc (`referrer` took precedence. Now only directs to referrer if grade and year already have values).
* Updates CSS and markup for banner component to match that used on line service (including ability to display an icon)
* Uses banner component to show blue/red incomplete warnings on review page (better reflecting component used on live service)
* Updates section review includes to display new incomplete banners
* Updates incomplete messages to match those used on live service
* Updates routes for review pages to validate every section (and moves this logic into a separate file)
* When creating a new `application` object, populates it with empty objects and `null` values

adamsilver commented on app/routes/application.js:
```diff
@@ -10,7 +10,25 @@ function createNewApplication (req) {
   if (typeof data.applications[code] === 'undefined') {
     data.applications[code] = {
-      status: 'started'
+      status: 'started',
```
It seems to me that you've added these defaults so that later, in the routes, you can check to see if the user added choices via the following code:

`if (Object.keys(choices).length === 0)`

I don't think we should define empty objects here. Instead I'd make the check:

`if(application.choices)`

If the fact it's empty is a little more complicated (which I think it might be judging from the fact that some properties are set to an empty object and some are set to null) then I'd suggest a method to encapsulate this so I can do:

`if(app.isTaskComplete('choices'))`

Behind the scenes ,this function might hand off to `app.isChoicesComplete()`. 

`

paulrobertlloyd replied:
Created a new utility function `hasCompletedSection(key)`, which checks for either `null` or empty keys.

adamsilver commented on app/routes/application/review.js:
```diff
@@ -0,0 +1,118 @@
+
+    if (successFlash[0] === 'submitted-incompleted-application') {
+      pageObject.errorList = []
+      const sections = {
```
This now looks like if the user submits an incomplete application then every single thing is marked as incomplete instead of just the incomplete things.

paulrobertlloyd replied:
With the changes made so far, behaviour appears correct; the errors shown at the top of the page reflect the incomplete sections. So may skip refactoring this.

adamsilver commented on app/routes/application/review.js:
```diff
@@ -0,0 +1,118 @@
+    // just checking choices are empty for now
+    // to make real need to check over every section
+    if (
+      Object.keys(choices).length === 0 ||
```
Instead of all these checks, I think this should be a method on the application like this:

`if(!app.isComplete())`

paulrobertlloyd replied:
Created a new utility function `hasCompletedApplication(req)`, which follows a similar design to the other checks we have.

adamsilver commented on app/routes/application/review.js:
```diff
@@ -0,0 +1,118 @@
+      req.flash('success', 'submitted-incompleted-application')
+      res.redirect(`/application/${req.params.applicationId}/review`)
+    } else {
+      console.log('applicationData.status', applicationData.status)
```
Rogue log

adamsilver commented on app/views/_components/banner/template.njk:
```diff
@@ -1,14 +1,4 @@
 <div class="app-banner{% if params.type == "success" %} app-banner--success{% elif params.type == "warning" %} app-banner--warning{% endif %}{{- " " + params.classes if params.classes}}" {%- for attribute, value in params.attributes -%} {{ attribute }}="{{ value }}"{% endfor %}>
-{% if params.icon != false %}
```
I think the icons are valuable and that we should show them in order to provide a comparable experience to users who don't see colour so well.

paulrobertlloyd replied:
Have restored the ability to include icons with the banner component, but disabled to main parity with what’s shipped (not adding/changing features in this PR, just making things work!)

paulrobertlloyd commented:
@adamsilver If you could take another look over this, that’d be grand. Also fixed an issue where trying to complete an incomplete GCSE or Degree from the review page, wouldn’t allow you to answer the follow on questions about year, grade etc. Think I’ve fixed that!

fofr commented:
Nice work! A much needed update to the prototype.

paulrobertlloyd commented:
@adamsilver @fofr Thanks both — this thing is a monster! 🤪
